synapse/handlers/federation_event.py

```diff
@@ -765,6 +766,11 @@ async def _process_pulled_event(
         """
         logger.info("Processing pulled event %s", event)
 
+        # TODO: Why does this matter? The whole point of this function is to
+        # persist random PDU's from backfill. It shouldn't matter whether we saw
+        # them somewhere else first as an outlier, then during backfill. This
+        # function handles de-outliering anyway.
```

Comment:
It seems like the only blocker here is the `assert not outlier` in a few places and that we `compute_event_context` which only works for non-outliers. The silly part is that raw PDU's over federation aren't `outliers` which is why normal backfill just works. But using the cached getters sometimes gives us an `outlier`. Seems like we should just be converting every event to a non-outlier in `_process_pulled_events` 🤷 


```
_process_pulled_events
_process_pulled_event
_process_received_pdu
compute_event_context
```

Reply:
(I'm looking for your thoughts here)

Reply:
> > TODO: Why does this matter?
>
> well, this function performs a bunch of operations that are not appropriate when we persist an outlier (in particular, it makes sure we have all the `prev_events` and resolves the state across those prev events).

Ahh, so it's more about doing the expected operations for persisting an outlier rather than we can't handle it (because we can). A safe-guard in case someone in the future refactors Synapse to persist outliers using this function (i.e. they should use something else).

In our case, we're backfilling an event so this usage is appropriate.

In that case, this needs a good comment and a good assertion message. Something like: "This event is marked as an outlier. If you're trying to persist an outlier, you should use a different function. If you're trying to backfill an event, this is the right method but you need pass in an event copy that doesn't have `event.internal_metada.outlier = true`."

Reply:
Updated 👍 I'm sure some language changes are needed

https://github.com/matrix-org/synapse/blob/337d8be9ea5fe10a563f0293e466eb4b79403477/synapse/handlers/federation_event.py#L769-L784